### PR TITLE
support json() +with pipeline in asyncio.cluster mode

### PR DESCRIPTION
**Pull request by nlsj1985**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

_Please provide a description of the change here._

Hi, as newbie redis-user I ran into #2234 with asyncio cluster support for json() + also in combination with pipelines
I've created some patches, that resulted in functioning support for this.
As i'm not practiced enough with big projects like these, I would like to request someone to review the changes that I've made. I might have broken some other things (or not), and I'm sure there are some rookie mistakes in the changes.. but it works for me.. and I've spent quite some hours on this.
So I hope someone who's more experienced could help pickup my improvements for the project, and improve upon it.
I'm guessing there also needs a bit of testing to get extended.
I unfortunately have no resources to complete that, and run the complete set op tests, and to investigate the results.
I've ran my test.py (see examples/) using python3.10.4 and the docker image redis-modcluster and that works fine for my usecase (looking from the surface, in my brief tests), I unfortunately have to carry-on doing other stuff.
Thanks in advance!

### CODE REVIEW

File: redis/asyncio/cluster.py [hunk 1]
```diff
@@ -152,6 +153,7 @@ class RedisCluster(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterCommand
         - none of the `host`/`port` & `startup_nodes` were provided
 
     """
+    response_callbacks: MutableMapping[Union[str, bytes], ResponseCallbackT]
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:156]: This is defined but never initialized.

File: redis/asyncio/cluster.py [hunk 2]
```diff
@@ -298,7 +300,10 @@ def __init__(
             # Call our on_connect function to configure READONLY mode
             kwargs["redis_connect_func"] = self.on_connect
 
-        kwargs["response_callbacks"] = self.__class__.RESPONSE_CALLBACKS.copy()
+        kwargs["cluster_response_callbacks"] = CaseInsensitiveDict(self.__class__.RESPONSE_CALLBACKS)
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:303]: What is the purpose of renaming this variable? Also, what is the advantage of using CaseInsensitiveDict? It is significantly slower that a normal dict due to the case conversion for each call.

File: redis/asyncio/cluster.py [hunk 3]
```diff
@@ -1206,7 +1211,7 @@ async def close(self, attr: str = "nodes_cache") -> None:
         )
 
 
-class ClusterPipeline(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterCommands):
+class ClusterPipeline(RedisCluster):
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:1214]: Why are the bases for this class changed? Inheriting directly from the RedisCluster class could lead to collisions in the namespace

File: redis/asyncio/cluster.py [hunk 4]
```diff
@@ -1245,9 +1250,21 @@ class ClusterPipeline(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterComm
 
     __slots__ = ("_command_stack", "_client")
 
-    def __init__(self, client: RedisCluster) -> None:
+    def __init__(
+            self,
+            client: RedisCluster,
+            nodes_manager=None,
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:1256]: The new kwargs are not used anywhere

File: redis/asyncio/cluster.py [hunk 5]
```diff
@@ -298,7 +300,10 @@ def __init__(
             # Call our on_connect function to configure READONLY mode
             kwargs["redis_connect_func"] = self.on_connect
 
-        kwargs["response_callbacks"] = self.__class__.RESPONSE_CALLBACKS.copy()
+        kwargs["cluster_response_callbacks"] = CaseInsensitiveDict(self.__class__.RESPONSE_CALLBACKS)
+        self.cluster_response_callbacks = CaseInsensitiveDict(
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:304]: This will always get overwritten on line 332.

File: redis/asyncio/cluster.py [hunk 6]
```diff
@@ -1245,9 +1250,21 @@ class ClusterPipeline(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterComm
...
+            startup_nodes=None,
+            read_from_replicas=False,
+            cluster_error_retry_attempts=5,
+            reinitialize_steps=10,
+            lock=None,
+            **kwargs,
+    ) -> None:
         self._client = client
-
+        self.cluster_response_callbacks = self.RESPONSE_CALLBACKS
```
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:1267]: This is not used anywhere?

File: redis/asyncio/cluster.py (hunk 1 above)
Comment (utkarshgupta137) [inline redis/asyncio/cluster.py:156]: This is defined but never initialized?

**Review by utkarshgupta137 — CHANGES_REQUESTED** [review]
I think you don't need any changes to the cluster.py except for adding AsyncRedisModuleCommands inheritance to the RedisCluster & ClusterPipeline classes.

**Comment by nlsj1985** [comment]
@utkarshgupta137 I was hopping you would check this out and provide pointers or improvements. As said, i'm really inexperienced with all this and hoped to help out.. but i was really struggling by comparing sync cluster, pipeline and json() with the current implemented async code, so just started copying stuff, run it, and patching the next exception in any way I could think of, till I got it working and then I was drained and confused by how things got to the point where it started to work ;-)

**Comment by nlsj1985** [comment]
I've reverted quite some changes and took your recommendations
cluster.py:
class RedisCluster(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterCommands, AsyncRedisModuleCommands):
class ClusterPipeline(AbstractRedis, AbstractRedisCluster, AsyncRedisClusterCommands, AsyncRedisModuleCommands):
    def __init__(self, client: RedisCluster) -> None:
        super().__init__()  # because my IDE say __init__ needed the super class initialized 
        self._client = client
        self._command_stack: List["PipelineCommand"] = []

resulting in:
  File "C:\Users\janse499\PycharmProjects\redis-py\redis\commands\redismodules.py", line 110, in json
    jj = AsyncJSON(client=self, encoder=encoder, decoder=decoder)
  File "C:\Users\janse499\PycharmProjects\redis-py\redis\commands\json\__init__.py", line 65, in __init__
    self.client.set_response_callback(key, value)
AttributeError: 'ClusterPipeline' object has no attribute 'set_response_callback'

set_response_callback() was used before from inheriting RedisCluster.. 
I've been looking a bit again over client.py and cluster.py and the different pipelines, and trying some stuff.. but this is to complex for me to fix at this moment. I hope you might have some more pointers, I'm happy to learn.

ps. also, when I removed RedisCluster from ClusterPipeline's inheritance the json\__init__.py is not accepting the parameters in AsyncioClusterPipeline() anymore at line 178

KR

**Comment by utkarshgupta137** [comment]
I've never used any redis modules, so I can't work on this myself. I can review cluster related code though. For now, I would recommend not changing anything in the cluster.py file except adding inheritance for async redis modules command to the client & pipeline classes.

**Pull request closed without merging**